README.md
# Very elegant, fast and powerful router for PHP

[![Latest Stable Version](https://img.shields.io/packagist/v/vitorsreis/extend-router?style=flat-square&label=stable&color=2E9DD3)](https://packagist.org/packages/vitorsreis/extend-router)
[![PHP Version Require](https://img.shields.io/packagist/dependency-v/vitorsreis/extend-router/php?style=flat-square&color=777BB3)](https://packagist.org/packages/vitorsreis/extend-router)
[![License](https://img.shields.io/packagist/l/vitorsreis/extend-router?style=flat-square&color=418677)](https://github.com/vitorsreis/extend-router/blob/master/LICENSE)
[![Total Downloads](https://img.shields.io/packagist/dt/vitorsreis/extend-router?style=flat-square&color=0476B7)](https://packagist.org/packages/vitorsreis/extend-router)
[![Repo Stars](https://img.shields.io/github/stars/vitorsreis/extend-router?style=social)](https://github.com/vitorsreis/extend-router)

Indexed by words tree and regex matching, this router is very elegant, fast, and powerful. Architected as a queue of
merged middlewares (not unique match), it offers multiple interactions in routes with cache, contexts, and persistent
data.
Unit tests passed on versions: ```5.6```, ```7.4```, ```8.1``` and ```8.2```

---

## Benchmark

Check out benchmark with leading public libraries [here](/tests/Benchmark/Benchmark.md).

---

## Install

```shell
composer require vitorsreis/extend-router
```

## Usage

```php
$router = new \VSR\Extend\Router();
$router->get('/product/:id', function ($id) { echo "product $id"; });
$router->match('GET', '/product/100')->execute();
// output: "product 100"
```

---

### Router group

You can add group of routes:

```php
$router->group('/product', function (\VSR\Extend\Router $router) {
    $router->get('/:id', function ($id) { echo "get product $id"; });
    $router->post('/:id', function ($id) { echo "post product $id"; });
});
$router->match('POST', '/product/100')->execute();
// output: "post product 100"
```

---

### 404 Not Found and 405 Method Not Allowed

- You can use NotFoundException and MethodNotAllowedException in catch:

```php
try {
    $router->match('POST', '/aaa');
} catch (\VSR\Extend\Router\Exception\NotFoundException $e) {
    echo "{$e->getCode()}: {$e->getMessage()}"; // output: "404: Route \"/aaa\" not found"
} catch (\VSR\Extend\Router\Exception\MethodNotAllowedException $e) {
    echo "{$e->getCode()}: {$e->getMessage()}"; // output: "405: Method \"POST\" not allowed for route \"/aaa\""
}
```

- Or can use RuntimeException in catch:

```php
try {
    $router->match('POST', '/aaa');
} catch (\VSR\Extend\Router\Exception\RuntimeException $e) {
    if (in_array($e->getCode(), [404, 405])) {
        echo "{$e->getCode()}: {$e->getMessage()}";
        // output: "404: Route \"/aaa\" not found" or  "405: Method \"POST\" not allowed for route \"/aaa\""
    } else {
        throw $e;
    }
}
```

- In case 405, you can get allowed methods:

```php
try {
    $router->match('PUT', '/aaa');
} catch (\VSR\Extend\Router\Exception\MethodNotAllowedException $e) {
    echo implode(', ', $e->allowedMethods); // output: "GET, POST, ..."
}
```

---

### Friendly URIs

You can add friendly URIs to redirect to specific routes:

```php
$router->post('/product/:id', function ($id) { echo "post product $id"; });
$router->friendly('/iphone', '/product/100');
$router->match('POST', '/iphone')->execute();
// output: "post product 100"
```

---

### Filters

Filters are used to add regex to route variables in a nicer and cleaner way.

```php
$router->get('/:var1[09]', function ($var1) { return "[09] : $var1"; });
$router->get('/:var1[az]', function ($var1) { return "[az] : $var1"; });

echo $router->match('GET', '/111')->execute()->result; // output: "[09] : 111"
echo $router->match('GET', '/aaa')->execute()->result; // output: "[az] : aaa"
```

You can use loose filter in routes:

```php
$router->get('/[09]', function () { return "[09]"; });
$router->get('/[az]', function () { return "[az]"; });

echo $router->match('GET', '/111')->execute()->result; // output: "[09]"
echo $router->match('GET', '/aaa')->execute()->result; // output: "[az]"
```

You can add custom filters:

```php
$router->addFilter('custom_only_numeric', '\d+')
$router->get('/:var1[custom_only_numeric]', function ($var1) { return "CUSTOM_VAR1_FILTER : $var1"; });

$router->addFilter('custom_w10', '\w{10}')
$router->get('/[custom_w10]', function () { return 'CUSTOM_LOOSE_FILTER'; });
```

Below are pre-registered filters:

|            Key             |         Description         | Regex                                                                           |
|:--------------------------:|:---------------------------:|---------------------------------------------------------------------------------|
| _&lt;omitted or empty&gt;_ |     any other than "/"      | \[^\/]+                                                                         |
|             09             |        only numbers         | \[0-9]+                                                                         |
|             az             |       only lowercase        | \[a-z]+                                                                         |
|             AZ             |       only uppercase        | \[A-Z]+                                                                         |
|             aZ             |        only letters         | \[a-zA-Z]+                                                                      |
|             d              |        only numbers         | \d+                                                                             |
|             D              |    any other than "0-9"     | \D+                                                                             | 
|             w              |      only "a-zA-Z0-9_"      | \w+                                                                             |
|             W              | any other than "a-zA-Z0-9_" | \W+                                                                             |
|            uuid            |            uuid             | \[0-9a-f]{8}-\[0-9a-f]{4}-\[0-5]\[0-9a-f]{3}-\[089ab]\[0-9a-f]{3}-\[0-9a-f]{12} |

---

### Allowing/Disallowing methods
Default list: ```GET```, ```POST```, ```PUT```, ```PATCH```, ```DELETE```, ```OPTIONS```, ```HEAD```

```php
$router->allowMethod('XXX', ...);
// $router->addRoute('XXX', '/aaa', function () { ... }); = OK
// $router->match('XXX', '/aaa'); = OK

$router->disallowMethod('GET', ...);
// $router->addRoute('GET', '/aaa', function () { ... }); = throw SyntaxException(500)
// $router->match('GET', '/aaa'); = throw RuntimeException(400)
```

---

### Cache

- This usage mode or ```Memory``` cache does not store the router map.

```php
$cache = new \VSR\Extend\Router\Cache\Memory();
// $cache = new \VSR\Extend\Router\Cache\File(__DIR__ . "/cache/");
// $cache = new \VSR\Extend\Router\Cache\Apcu();
// $cache = new \VSR\Extend\Router\Cache\Memcache();
// $cache = new \VSR\Extend\Router\Cache\Memcached();
// $cache = new \VSR\Extend\Router\Cache\Redis();

$router = new \VSR\Extend\Router($cache);
$router->get('/product/:id', function ($id) { echo "product $id"; });
$router->friendly('/iphone-xs', '/product/100');
$router->friendly('/samsumg-s10', '/product/200');
$router->match('GET', '/iphone-xs')->execute();
// output: "product 100"
```

- Use like this to instantiate already stored router or, if it doesn't exist, instantiate it and then store it in cache.
  This greatly reduces time from second load onwards. **Does not support routes with anonymous classes or
  anonymous/arrow functions (Closure objects)**
    - ```&$hash``` argument can be used to control the cache version, if omitted, it is based on the ```$callback```
      source code by reflection.
    - ```&$warning``` argument if different from ```false```", indicates that an error occurred.

```php
$cache = new \VSR\Extend\Router\Cache\File(__DIR__ . "/cache/");
// $cache = new \VSR\Extend\Router\Cache\Apcu();
// $cache = new \VSR\Extend\Router\Cache\Memcache();
// $cache = new \VSR\Extend\Router\Cache\Memcached();
// $cache = new \VSR\Extend\Router\Cache\Redis();

function callback($id) {
    echo "product $id";
}

$router = $cache->createRouter(function (\VSR\Extend\Router $router) {
    $router->get('/product/:id', 'callback');
    $router->friendly('/iphone-xs', '/product/100');
    $router->friendly('/samsumg-s10', '/product/200');
}, $hash, $warning);
$router->match('GET', '/iphone-xs')->execute();
// output: "product 100"
```

- Allow\Disallow cache prefixes for store.
    - ```FLAG_ROUTER``` to enable/disable cache router
    - ```FLAG_MATCH``` to enable/disable cache match
    - ```FLAG_EXECUTE``` to enable/disable cache execute
    - ```FLAG_OTHERS``` to enable/disable cache others
    - ```FLAG_ALL``` to enable/disable cache all

```php
$cache = new \VSR\Extend\Router\Cache\Memory();
$cache->disallowCache($cache::FLAG_ALL);
$cache->allowCache($cache::FLAG_ROUTER);
```

---

### Context param

Context contains all information about the current execution. Use an argument with the name "$context" of type "
omitted,""mixed," or "\VSR\Extend\Router\Context" in middlewares or in the constructor of a class if the middleware is
of type non-static class method

```php
$router->get('/aaa', function ($context) { ... });
$router->any('/aaa', function (mixed $context) { ... }); # Explicit mixed type only PHP 8+
$router->get('/a*', function (\VSR\Extend\Router\Context $context) { ... });
$router->any('/a*', function (\VSR\Extend\Router\Context $custom_name_context) { ... });
```

---

### Context methods/properties

| Property                                   | Description                                                                        |
|:-------------------------------------------|:-----------------------------------------------------------------------------------|
| ```$context->current->route```             | Current match middleware route                                                     |
| ```$context->current->method```            | Current match middleware method                                                    |
| ```$context->current->uri```               | Current match middleware uri                                                       |
| ```$context->current->friendly```          | Current match middleware friendly uri                                              |
| ```$context->current->params```            | Current match middleware uri variables                                             |
| ```$context->header->hash```               | Current execution hash                                                             |
| ```$context->header->cursor```             | Current execution position on queue middleware                                     |
| ```$context->header->total```              | Total execution queue middlewares count                                            |
| ```$context->header->state```              | Current execution state                                                            |
| ```$context->header->startTime```          | Execution start time                                                               |
| ```$context->header->endTime```            | Execution end time                                                                 |
| ```$context->header->elapsedTime```        | Execution time                                                                     |
| ```$context->cached```                     | Execution result is cached                                                         |
| ```$context->allowedMethods```             | Current match allowed methods                                                      |
| ```$context->result```                     | Partial/Final execution result                                                     |
| ```$context->execute(?$callback)```        | Start execution, ```$callback``` is optional and available argument ```$context``` |
| ```$context->stop()```                     | Stop execution                                                                     |
| ```$context->get($key, $default = null)``` | Get persistent data                                                                |
| ```$context->set($key, $value)```          | Set persistent data                                                                |
| ```$context->has($key)```                  | Check if persistent data exists                                                    |

---

### Persisting data

You can persist data in context so that it is persisted in future callbacks.

```php
$router->get('/aaa', function (\VSR\Extend\Router\Context $context) {
    $context->set('xxx', $context->get('xxx', 0) + 10); # 2. Increment value: 5 + 10 = 15
});
$router->get('/var2', function (\VSR\Extend\Router\Context $context) {
    $context->set('xxx', $context->get('xxx', 0) + 15); # 3. Increment value: 15 + 15 = 30
});
$context = $router->match('GET', '/aaa')
    ->set('xxx', 5) # 1. Initial value: 5
    ->execute();

echo $context->get('xxx');
// output: "30"
```

---

### Execute with callback

You can run receiving callbacks every middleware run with current context.

```php
$router->post('/:aa', function ($aa) { return "a1:$aa "; }, function ($aa) { return "a2:$aa "; });
$router->post('/:bb', function ($bb) { return "bb:$bb "; });
$router->match('POST', '/99')->execute(function ($context) {
    // partial result, run 3 times
    echo '[' .
        "{$context->header->cursor}/{$context->header->total} " .
        "{$context->current->method} {$context->current->route} = $context->result" .
    '], ';
});
// output: "[1/3 POST /:aa = a1:99], [2/3 POST /:aa = a2:99], [3/3 POST /:bb = bb:99], "
```

---

### Middleware with arguments

Route variables and context param are not mandatory in callbacks, so they can be omitted without problems.

```php
$router->any('/:var1/:var2', function () { ... });
$router->any('/:var1/:var2', function ($var1) { ... });
$router->any('/:var1/:var2', function ($var2) { ... });
$router->any('/:var1/:var2', function ($context) { ... });
$router->any('/:var1/:var2', function (mixed $context) { ... }); # Explicit mixed type only PHP 8+
$router->any('/:var1/:var2', function (\VSR\Extend\Router\Context $context) { ... });
$router->any('/:var1/:var2', function (\VSR\Extend\Router\Context $custom_name_context) { ... });
$router->any('/:var1/:var2', function ($var1, $var2) { ... });
$router->any('/:var1/:var2', function ($var1, $context) { ... });
$router->any('/:var1/:var2', function ($var1, mixed $context) { ... }); # Explicit mixed type only PHP 8+
$router->any('/:var1/:var2', function ($var1, \VSR\Extend\Router\Context $context) { ... });
$router->any('/:var1/:var2', function ($var1, \VSR\Extend\Router\Context $custom_name_context) { ... });
$router->any('/:var1/:var2', function ($var2, $context) { ... });
$router->any('/:var1/:var2', function ($var2, mixed $custom_name_context) { ... }); # Explicit mixed type only PHP 8+
$router->any('/:var1/:var2', function ($var2, \VSR\Extend\Router\Context $context) { ... });
$router->any('/:var1/:var2', function ($var2, \VSR\Extend\Router\Context $custom_name_context) { ... });
$router->any('/:var1/:var2', function ($var1, $var2, $context) { ... });
$router->any('/:var1/:var2', function ($var1, $var2, mixed $context) { ... }); # Explicit mixed type only PHP 8+
$router->any('/:var1/:var2', function ($var1, $var2, \VSR\Extend\Router\Context $context) { ... });
$router->any('/:var1/:var2', function ($var1, $var2, \VSR\Extend\Router\Context $custom_name_context) { ... });
```

---

### Middleware queue

With the "not unique match" pattern, you can have multiple callbacks in queue per order of addition for an uri.

```php
$router->get('/aaa', function () { echo "11 "; }, function () { echo "12 "; }, function () { echo "13 "; });
$router->any('/aaa', function () { echo "2 "; });
$router->get('/a*', function () { echo "3 "; });
$router->any('/a*', function () { echo "4 "; });
$router->get('*', function () { echo "5 "; });
$router->any('*', function () { echo "6 "; });
$router->get('/:var', function ($var) { echo "7 "; });
$router->any('/:var', function ($var) { echo "8 "; });
$router->match('GET', '/aaa')->execute();
// output: "11 12 13 2 3 4 5 6 7 8 "
```

You can stop the queue using "stop" method of context

```php
$router->get('/aaa', function () { echo "11 "; }, function () { echo "12 "; }, function () { echo "13 "; });
$router->any('/aaa', function () { echo "2 "; });
$router->get('/a*', function () { echo "3 "; });
$router->any('/a*', function (\VSR\Extend\Router\Context $context) { echo "4 "; $context->stop(); });
$router->get('*', function () { echo "5 "; });
$router->any('*', function () { echo "6 "; });
$router->get('/:var', function ($var) { echo "7 "; });
$router->any('/:var', function ($var) { echo "8 "; });
$router->match('GET', '/aaa')->execute();
// output: "11 12 13 2 3 4 "
```

---

### Supported middleware types

```php
// by native function name
$router->any('/:haystack/:needle', "stripos");

#--------------------------------------------------

// by function name
function callback($var1, $var2, $context) { ... }
$router->any('/:var1/:var2', "callback");

#--------------------------------------------------

// by anonymous function
$router->any('/:var1/:var2', function ($var1, $var2, $context) { ... });
$router->any('/:var1/:var2', static function ($var1, $var2, $context) { ... });

#--------------------------------------------------

// by arrow function, PHP 7.4+
$router->any('/:var1/:var2', fn($var1, $var2, $context) => ...);
$router->any('/:var1/:var2', static fn($var1, $var2, $context) => ...);

#--------------------------------------------------

// by variable function
$callback1 = function ($var1, $var2, $context) { ... };
$router->any('/:var1/:var2', $callback1);

$callback2 = static function ($var1, $var2, $context) { ... };
$router->any('/:var1/:var2', $callback2);

#--------------------------------------------------

// by class method
class AAA {
    public function method($var1, $var2, $context) { ... }
}
$aaa = new AAA();

$router->any('/:var1/:var2', "AAA::method"); // Call first constructor if exists and then method
$router->any('/:var1/:var2', [ AAA::class, 'method' ]); // Call first constructor if exists and then method
$router->any('/:var1/:var2', [ new AAA(), 'method' ]); // Call method
$router->any('/:var1/:var2', [ $aaa, 'method' ]); // Call method

#--------------------------------------------------

// by class static method
class BBB {
    public static function method($var1, $var2, $context) { ... }
}
$bbb = new BBB();

$router->any('/:var1/:var2', "BBB::method"); // Call static method
$router->any('/:var1/:var2', [ BBB::class, 'method' ]); // Call static method
$router->any('/:var1/:var2', [ new BBB(), 'method' ]); // Call static method
$router->any('/:var1/:var2', [ $bbb, 'method' ]); // Call static method

#--------------------------------------------------

// by class method with constructor
class CCC {
    public function __construct($context) { ... }
    public function method($var1, $var2, $context) { ... }
}

$router->any('/:var1/:var2', "CCC::method"); // Call first constructor and then method
$router->any('/:var1/:var2', [ CCC::class, "method" ]); // Call first constructor and then method

#--------------------------------------------------

// by class name/object
class DDD {
    public function __invoke($var1, $var2, $context) { ... }
}
$ddd = new DDD();

$router->any('/:var1/:var2', "DDD"); // Call first constructor if exists and then __invoke
$router->any('/:var1/:var2', DDD::class); // Call first constructor if exists and then __invoke
$router->any('/:var1/:var2', new DDD()); // Call __invoke
$router->any('/:var1/:var2', $ddd); // Call __invoke

#--------------------------------------------------

// by anonymous class, PHP 7+
$router->any('/:var1/:var2', new class {
    public function __invoke($var1, $var2, $context) { ... }
}); // Call __invoke
```

CHANGELOG.md
CHANGELOG
---

## 4.5.0
- Add create custom params on execute

## 4.4.2
- Fix joker empty match

## 4.4.1
- Fix class inheritance

## 4.4.0
- Implement group without callback
- Add methods allowMethod/disallowMethod

## 4.3.1
- Fix deprecated alert matcher sort
- Add property allowedMethods in Context and MethodNotAllowedException

## 4.3.0
- Use [VSR\\Extend\\Caller](https://github.com/vitorsreis/extend-caller)
- Cache improvements
  - Add permission control for cache prefixes
    - flags: FLAG_ROUTER, FLAG_MATCH, FLAG_EXECUTE, FLAG_OTHERS, FLAG_ALL
    - allowCache(flag)
    - disallowCache(flag)
- Fix property cached in match result
- Fix PSR-12

## 4.2.0
- Added router map builder/store in cache
- Cache improvements
  - Fix get File Cache

## 4.1.1
- Fix variable params
- Fix add filter
- Performance improvements
  - Fix unique added route index in route collection variable tree
  - Fix match indexes tree
  - Fix match middlewares order
- Cache improvements
  - Fix cache get value/default

## 4.1.0
- Add router group
- Cache improvements
  - Add execution hash ($context->header->hash)
  - Add clear() method
  - Add match result to cache
  - Add execution result to cache
  - Add APCU cache
  - Add File cache
  - Add Redis cache
  - Add Memcache cache
  - Add Memcached cache
- Fix filter parser delimiter from / to ~
- Add exception NotFoundException and MethodNotAllowedException

## 4.0.0
- Downgrade for PHP 5.6+ usage
- PSR-12 code style
- Add throw reserved name "context" from router
- Add callback on context execute
- Fix caller optional params
- Fix cursor position

## 3.0.2
- Performance improvements
- Fix benchmark

## 3.0.1
- Change filter syntax: **var|filter** to **var\[filter]**
- Use loose filter in routes **\[filter]**
- Fix params mapping
- Fix static router indexer

## 3.0.0
- Full restructuring


src/Manager/Parser.php
<?php

/**
 * This file is part of d5whub extend router
 * @author Vitor Reis <vitor@d5w.com.br>
 */

namespace D5WHUB\Extend\Router\Manager;

use Closure;
use D5WHUB\Extend\Router\Context;
use D5WHUB\Extend\Router\Exception\RuntimeException;
use D5WHUB\Extend\Router\Exception\SyntaxException;
use ReflectionException;
use ReflectionFunction;
use ReflectionFunctionAbstract;
use ReflectionMethod;

trait Parser
{
    /**
     * @param string|string[] $httpMethods
     * @param int $httpCode
     * @return string[]
     * @throws RuntimeException
     * @throws SyntaxException
     */
    private function parseHttpMethods($httpMethods, $httpCode = 500)
    {
        if (is_string($httpMethods)) {
            $httpMethods = [$httpMethods];
        }

        return array_map(
            static function ($httpMethod) use ($httpCode) {
                switch ($httpMethod) {
                    case 'ANY':
                    case 'GET':
                    case 'POST':
                    case 'PUT':
                    case 'PATCH':
                    case 'DELETE':
                    case 'OPTIONS':
                    case 'HEAD':
                        return $httpMethod;

                    default:
                        throw $httpCode === 400
                            ? new RuntimeException("Http method \"$httpMethod\" invalid", $httpCode)
                            : new SyntaxException("Http method \"$httpMethod\" invalid", $httpCode);
                }
            },
            $httpMethods
        );
    }

    /**
     * @param string $key
     * @param string $pattern
     * @return array{string, string}
     * @throws SyntaxException
     */
    private function parseFilter($key, $pattern)
    {
        if (preg_match(self::REGEX_FILTER_KEY, $key)) {
            throw new SyntaxException("Invalid key \"$key\". Use only letters, numbers and underscore", 500);
        }

        $try = self::REGEX_DELIMITER . $pattern . self::REGEX_DELIMITER;
        if (!@preg_match($try, '') && preg_last_error() !== PREG_NO_ERROR) {
            throw new SyntaxException("Invalid pattern \"$try\" on filter key \"$key\"", 500);
        }

        return [$key, $pattern];
    }

    /**
     * @param string $uri
     * @return string
     */
    private function parseUri($uri)
    {
        do {
            $uri = preg_replace(self::REGEX_DELIMITER . '\*\*' . self::REGEX_DELIMITER, '*', $uri, -1, $count);
        } while ($count);

        do {
            $uri = preg_replace(self::REGEX_DELIMITER . '//' . self::REGEX_DELIMITER, '/', $uri, -1, $count);
        } while ($count);

        return rtrim(parse_url($uri, PHP_URL_PATH), '/') ?: '/';
    }

    /**
     * @param string $route
     * @return array{string, string, string[], string[], string[]}
     * @throws SyntaxException
     */
    private function parseRoute($route)
    {
        $route = $this->parseUri($route);

        $static = true;
        $pattern = '';
        $words = [];
        $paramNames = [];
        $argCount = 0;

        if ($split = preg_split(self::REGEX_MATCH, $route, -1, self::PREG_SPLIT_FLAGS)) {
            foreach ($split as $match) {
                if ($match[0] === ':') {
                    $words[] = ':';

                    $matchSplit = preg_split(self::REGEX_VARIABLE, $match, 2, self::PREG_SPLIT_FLAGS);
                    $paramName = $matchSplit[0];

                    if (strcasecmp($paramName, 'context') == 0) {
                        throw new SyntaxException("Param with reserved name \":$paramName\"", 500);
                    }

                    if (in_array($paramName, $paramNames)) {
                        throw new SyntaxException("Param with duplicate name \":$paramName\"", 500);
                    }

                    $filterKey = (isset($matchSplit[1]) ? $matchSplit[1] : '*') ?: '*';
                    if (!isset($this->filterCollection[$filterKey])) {
                        throw new SyntaxException("Filter \"$filterKey\" not implemented", 500);
                    }

                    $static = false;
                    $pattern .= "(?<A$argCount>" . $this->filterCollection[$filterKey] . ")";
                    $paramNames[$argCount] = $paramName;
                    $argCount++;
                } elseif ($match[0] === '[') {
                    $words[] = '*';

                    $matchSplit = preg_split(self::REGEX_LOOSE_FILTER, $match, 2, self::PREG_SPLIT_FLAGS);
                    $filterKey = isset($matchSplit[0]) ? $matchSplit[0] : '';

                    if (!isset($this->filterCollection[$filterKey])) {
                        throw new SyntaxException("Filter \"$filterKey\" not implemented", 500);
                    }

                    $static = false;
                    $pattern .= $this->filterCollection[$filterKey];
                } else {
                    $words[] = $match;

                    if ($match[0] === '*') {
                        $static = false;
                        $pattern .= '.*?';
                    } elseif ($match[0] === '/') {
                        $pattern .= '\/';
                    } else {
                        $pattern .= preg_quote($match, self::REGEX_DELIMITER);
                    }
                }
            }
        }

        return [$route, $pattern, $paramNames, $static, $words];
    }

    /**
     * @param array $middlewares
     * @return array{
     *     current: string,
     *     callback: callable,
     *     params: array<string, mixed>,
     *     construct: array<string, mixed>
     * }[]
     * @throws RuntimeException
     */
    private function parseMiddlewares(array $middlewares)
    {
        return array_map(
            function ($middleware) {
                $current = $middleware['current'];
                $callback = $middleware['callback'];

                if (is_object($callback) && !is_a($callback, Closure::class)) {
                    # anonymous class
                    $callback = [$callback, '__invoke'];
                } elseif (is_string($callback)) {
                    if (strpos($callback, '::') !== false) {
                        # class::method string
                        $callback = explode('::', $callback);
                    } elseif (!function_exists($callback) && class_exists($callback)) {
                        # class::__invoke
                        $callback = [$callback, '__invoke'];
                    }
                }

                if (is_array($callback)) {
                    # method / static method
                    try {
                        $reflection = new ReflectionMethod($callback[0], $callback[1]);
                    } catch (ReflectionException $e) {
                        throw new RuntimeException($e->getMessage(), 500, $e);
                    }

                    if (!$reflection->isStatic()) {
                        $construct = method_exists($callback[0], '__construct')
                            ? $this->parseMiddlewareParams(new ReflectionMethod($callback[0], '__construct'), true)
                            : [];

                        $callback = [
                            'callable' => $callback,
                            'params' => $this->parseMiddlewareParams($reflection),
                            'construct' => $construct
                        ];
                    } else {
                        $callback = [
                            'callable' => $callback,
                            'params' => $this->parseMiddlewareParams($reflection),
                            'construct' => null
                        ];
                    }
                } else {
                    # function / anonymous function / arrow function / string function
                    try {
                        $reflection = new ReflectionFunction($callback);
                        $callback = [
                            'callable' => $callback,
                            'params' => $this->parseMiddlewareParams($reflection),
                            'construct' => null
                        ];
                    } catch (ReflectionException $e) {
                        throw new RuntimeException($e->getMessage(), 500, $e);
                    }
                }

                return ['current' => $current] + $callback;
            },
            $middlewares
        );
    }

    /**
     * @param ReflectionFunctionAbstract $reflection
     * @param bool $onlyContext
     * @return array
     * @throws RuntimeException
     */
    private function parseMiddlewareParams($reflection, $onlyContext = false)
    {
        $result = [];

        foreach ($reflection->getParameters() as $param) {
            $paramType = method_exists($param, 'getType') ? $param->getType() : $param->getClass();
            if (!$paramType && $param->getName() === 'context') {
                # "$context"
                $result[$param->getName()] = ['type' => 'context', 'name' => $reflection->getName()];
                continue;
            }

            if ($paramType) {
                if (method_exists($paramType, 'getTypes')) {
                    $allowed = array_map(static function ($i) {
                        return $i->getName();
                    }, $paramType->getTypes());
                } elseif (method_exists($paramType, 'getName')) {
                    $allowed = [$paramType->getName()];
                } else {
                    $allowed = ['mixed'];
                }

                if (in_array('mixed', $allowed)) {
                    # "mixed $context"
                    $result[$param->getName()] = ['type' => 'context', 'name' => $reflection->getName()];
                    continue;
                }

                if (in_array(Context::class, $allowed)) {
                    # "Context $..."
                    $result[$param->getName()] = ['type' => 'context', 'name' => $reflection->getName()];
                    continue;
                }
            }

            if ($param->isOptional()) {
                $result[$param->getName()] = [
                    'type' => 'param',
                    'name' => $reflection->getName(),
                    'default' => $param->isDefaultValueAvailable() ? $param->getDefaultValue() : null
                ];
            } elseif ($onlyContext) {
                throw new RuntimeException(
                    sprintf("Required argument \"%s\" for invoke \"%s\"", $param->getName(), $reflection->getName()),
                    500
                );
            } else {
                $result[$param->getName()] = ['type' => 'param', 'name' => $reflection->getName()];
            }
        }

        return $result;
    }
}
